File: src/main/java/faba/java/Analysis.java
package faba.java;

import org.objectweb.asm.Opcodes;
import org.objectweb.asm.Type;
import org.objectweb.asm.tree.MethodNode;
import org.objectweb.asm.tree.analysis.AnalyzerException;
import org.objectweb.asm.tree.analysis.BasicValue;
import org.objectweb.asm.tree.analysis.Frame;

import java.util.*;
import static faba.java.AbstractValues.*;

class AbstractValues {
    static final class ParamValue extends BasicValue {
        ParamValue(Type tp) {
            super(tp);
        }
    }
    static final BasicValue InstanceOfCheckValue = new BasicValue(Type.INT_TYPE) {
        @Override
        public boolean equals(Object value) {
            return this == value;
        }
    };

    static final BasicValue TrueValue = new BasicValue(Type.INT_TYPE) {
        @Override
        public boolean equals(Object value) {
            return this == value;
        }
    };

    static final BasicValue FalseValue = new BasicValue(Type.INT_TYPE) {
        @Override
        public boolean equals(Object value) {
            return this == value;
        }
    };

    static final BasicValue NullValue = new BasicValue(Type.getObjectType("null")) {
        @Override
        public boolean equals(Object value) {
            return this == value;
        }
    };
    static final class NotNullValue extends BasicValue {
        NotNullValue(Type tp) {
            super(tp);
        }
    }
    static final class CallResultValue extends BasicValue {
        final Set<Key> inters;
        CallResultValue(Type tp, Set<Key> inters) {
            super(tp);
            this.inters = inters;
        }
    }

    static boolean isInstance(Conf curr, Conf prev) {
        if (curr.insnIndex != prev.insnIndex) {
            return false;
        }
        Frame<BasicValue> currFr = curr.frame;
        Frame<BasicValue> prevFr = prev.frame;
        for (int i = 0; i < currFr.getLocals(); i++) {
            if (!isInstance(currFr.getLocal(i), prevFr.getLocal(i))) {
                return false;
            }
        }
        for (int i = 0; i < currFr.getStackSize(); i++) {
            if (!isInstance(currFr.getStack(i), prevFr.getStack(i))) {
                return false;
            }
        }
        return true;
    }

    static boolean isInstance(BasicValue curr, BasicValue prev) {
        if (prev instanceof ParamValue) {
            return curr instanceof ParamValue;
        }
        if (InstanceOfCheckValue == prev) {
            return InstanceOfCheckValue == curr;
        }
        if (TrueValue == prev) {
            return TrueValue == curr;
        }
        if (FalseValue == prev) {
            return FalseValue == curr;
        }
        if (NullValue == prev) {
            return NullValue == curr;
        }
        if (prev instanceof NotNullValue) {
            return curr instanceof NotNullValue;
        }
        if (prev instanceof CallResultValue) {
            if (curr instanceof CallResultValue) {
                CallResultValue prevCall = (CallResultValue) prev;
                CallResultValue currCall = (CallResultValue) curr;
                return prevCall.inters.equals(currCall.inters);
            }
            else {
                return false;
            }
        }
        System.out.print("");
        return true;
    }
}

final class Conf {
    final int insnIndex;
    final Frame<BasicValue> frame;

    Conf(int insnIndex, Frame<BasicValue> frame) {
        this.insnIndex = insnIndex;
        this.frame = frame;
    }
}

final class State {
    final int index;
    final Conf conf;
    final List<Conf> history;
    final boolean taken;
    final boolean hasCompanions;
    final int insnIndex;

    State(int index, Conf conf, List<Conf> history, boolean taken, boolean hasCompanions) {
        this.index = index;
        this.conf = conf;
        this.history = history;
        this.taken = taken;
        this.hasCompanions = hasCompanions;
        insnIndex = conf.insnIndex;
    }
}

interface PendingAction<Res> {}
class ProceedState<Res> implements PendingAction<Res> {
    final State state;

    ProceedState(State state) {
        this.state = state;
    }
}
class MakeResult<Res> implements PendingAction<Res> {
    final State state;
    final Res subResult;
    final List<Integer> indices;

    // TODO - indices array
    MakeResult(State state, Res subResult, List<Integer> indices) {
        this.state = state;
        this.subResult = subResult;
        this.indices = indices;
    }
}

abstract class Analysis<Res> {
    final RichControlFlow richControlFlow;
    final Direction direction;
    final ControlFlowGraph controlFlow;
    final MethodNode methodNode;
    final Method method;
    final DFSTree dfsTree;
    final Res myIdentity;

    final Deque<PendingAction<Res>> pending = new LinkedList<PendingAction<Res>>();
    final Map<Integer, List<State>> computed = new HashMap<Integer, List<State>>();
    final Map<Integer, Res> results = new HashMap<Integer, Res>();
    final Key aKey;

    Res earlyResult = null;

    abstract Res identity();
    abstract Res combineResults(Res delta, List<Res> subResults);
    abstract boolean isEarlyResult(Res res);
    abstract Equation<Key, Value> mkEquation(Res result);
    abstract void processState(State state) throws AnalyzerException;

    protected Analysis(RichControlFlow richControlFlow, Direction direction) {
        this.richControlFlow = richControlFlow;
        this.direction = direction;
        controlFlow = richControlFlow.controlFlow;
        methodNode = controlFlow.methodNode;
        method = new Method(controlFlow.className, methodNode.name, methodNode.desc);
        dfsTree = richControlFlow.dfsTree;
        aKey = new Key(method, direction);
        myIdentity = identity();
    }

    final State createStartState() {
        return new State(0, new Conf(0, createStartFrame()), new ArrayList<Conf>(), false, false);
    }

    final boolean stateInstance(State curr, State prev) {
        if (curr.taken != prev.taken) {
            return false;
        }
        if (!isInstance(curr.conf, prev.conf)) {
            return false;
        }
        if (curr.history.size() != prev.history.size()) {
            return false;
        }
        for (int i = 0; i < curr.history.size(); i++) {
            if (!isInstance(curr.history.get(i), prev.history.get(i))) {
                return false;
            }
        }
        return true;
    }

    final Equation<Key, Value> analyze() throws AnalyzerException {
        pending.push(new ProceedState<Res>(createStartState()));
        while (!pending.isEmpty() && earlyResult == null) {
            PendingAction<Res> action = pending.pop();
            if (action instanceof MakeResult) {
                MakeResult<Res> makeResult = (MakeResult<Res>) action;
                ArrayList<Res> subResults = new ArrayList<Res>();
                for (int index : makeResult.indices) {
                    subResults.add(results.get(index));
                }
                Res result = combineResults(makeResult.subResult, subResults);
                if (isEarlyResult(result)) {
                    earlyResult = result;
                } else {
                    State state = makeResult.state;
                    int insnIndex = state.insnIndex;
                    results.put(state.index, result);
                    List<State> thisComputed = computed.get(insnIndex);
                    if (thisComputed == null) {
                        thisComputed = new ArrayList<State>();
                        computed.put(insnIndex, thisComputed);
                    }
                    thisComputed.add(state);
                }
            }
            else if (action instanceof ProceedState) {
                ProceedState<Res> proceedState = (ProceedState<Res>) action;
                State state = proceedState.state;
                int insnIndex = state.insnIndex;
                Conf conf = state.conf;
                List<Conf> history = state.history;

                boolean fold = false;
                if (dfsTree.loopEnters.contains(insnIndex)) {
                    for (Conf prev : history) {
                        if (isInstance(conf, prev)) {
                            fold = true;
                        }
                    }
                }
                if (fold) {
                    results.put(state.index, myIdentity);
                    List<State> thisComputed = computed.get(insnIndex);
                    if (thisComputed == null) {
                        thisComputed = new ArrayList<State>();
                        computed.put(insnIndex, thisComputed);
                    }
                    thisComputed.add(state);
                }
                else {
                    State baseState = null;
                    List<State> thisComputed = computed.get(insnIndex);
                    if (thisComputed != null) {
                        for (State prevState : thisComputed) {
                            if (stateInstance(state, prevState)) {
                                baseState = prevState;
                                break;
                            }
                        }
                    }
                    if (baseState != null) {
                        results.put(state.index, results.get(baseState.index));
                    } else {
                        // the main call
                        processState(state);
                    }

                }
            }
        }
        if (earlyResult != null) {
            return mkEquation(earlyResult);
        } else {
            return mkEquation(results.get(0));
        }
    }

    final Frame<BasicValue> createStartFrame() {
        Frame<BasicValue> frame = new Frame<BasicValue>(methodNode.maxLocals, methodNode.maxStack);
        Type returnType = Type.getReturnType(methodNode.desc);
        BasicValue returnValue = Type.VOID_TYPE.equals(returnType) ? null : new BasicValue(returnType);
        frame.setReturn(returnValue);

        Type[] args = Type.getArgumentTypes(methodNode.desc);
        int local = 0;
        if ((methodNode.access & Opcodes.ACC_STATIC) == 0) {
            frame.setLocal(local++, new BasicValue(Type.getObjectType(controlFlow.className)));
        }
        for (int i = 0; i < args.length; i++) {
            BasicValue value;
            if (direction instanceof InOut && ((InOut)direction).paramIndex == i) {
                value = new ParamValue(args[i]);
            }
            else if (direction instanceof In && ((In)direction).paramIndex == i) {
                value = new ParamValue(args[i]);
            }
            else {
                value = new BasicValue(args[i]);
            }
            frame.setLocal(local++, value);
            if (args[i].getSize() == 2) {
                frame.setLocal(local++, BasicValue.UNINITIALIZED_VALUE);
            }
        }
        while (local < methodNode.maxLocals) {
            frame.setLocal(local++, BasicValue.UNINITIALIZED_VALUE);
        }
        return frame;
    }

    final BasicValue popValue(Frame<BasicValue> frame) {
        return frame.getStack(frame.getStackSize() - 1);
    }

    final  <A> List<A> append(List<A> xs, A x) {
        ArrayList<A> result = new ArrayList<A>();
        if (xs != null) {
            result.addAll(xs);
        }
        result.add(x);
        return result;
    }
}

File: src/main/scala/analysis.scala
package faba.analysis

import scala.collection.mutable
import scala.collection.immutable.IntMap

import org.objectweb.asm.tree.analysis.{BasicValue, Frame}
import org.objectweb.asm.{Opcodes, Type}

import faba.cfg._
import faba.data._
import faba.engine._

case class ParamValue(tp: Type) extends BasicValue(tp)
case class InstanceOfCheckValue() extends BasicValue(Type.INT_TYPE)
case class TrueValue() extends BasicValue(Type.INT_TYPE)
case class FalseValue() extends BasicValue(Type.INT_TYPE)
case class NullValue() extends BasicValue(Type.getObjectType("null"))
case class NotNullValue(tp: Type) extends BasicValue(tp)
case class CallResultValue(tp: Type, inters: Set[Key]) extends BasicValue(tp)

case class Conf(insnIndex: Int, frame: Frame[BasicValue])

case class State(index: Int, conf: Conf, history: List[Conf], taken: Boolean, hasCompanions: Boolean) {
  val insnIndex: Int = conf.insnIndex
  override def toString = insnIndex.toString
}

object Counter {
  var processed: Long = 0
  var nonLocalDriving: Long = 0
  var nonShared: Long = 0
  var shared: Long = 0
  var effectivelyShared: Long = 0
}

abstract class Analysis[Res] {

  sealed trait PendingAction
  case class ProceedState(state: State) extends PendingAction
  case class MakeResult(state: State, subResult: Res, indices: List[Int]) extends PendingAction

  val richControlFlow: RichControlFlow
  val direction: Direction
  val stable: Boolean
  def identity: Res
  def processState(state: State): Unit
  def isEarlyResult(res: Res): Boolean
  def combineResults(delta: Res, subResults: List[Res]): Res
  def mkEquation(result: Res): Equation[Key, Value]

  val controlFlow = richControlFlow.controlFlow
  val methodNode = controlFlow.methodNode
  val method = Method(controlFlow.className, methodNode.name, methodNode.desc)
  val dfsTree = richControlFlow.dfsTree
  val aKey = Key(method, direction, stable)

  final def createStartState(): State = State(0, Conf(0, createStartFrame()), Nil, false, false)
  final def confInstance(curr: Conf, prev: Conf): Boolean = Utils.isInstance(curr, prev)

  final def stateEquiv(curr: State, prev: State): Boolean =
    curr.taken == prev.taken &&
      Utils.equiv(curr.conf, prev.conf) &&
      curr.history.size == prev.history.size &&
      (curr.history, prev.history).zipped.forall(Utils.equiv)

  val pending = mutable.Stack[PendingAction]()
  // the key is insnIndex
  var computed = IntMap[List[State]]().withDefaultValue(Nil)
  // the key is stateIndex
  var results = IntMap[Res]()

  var earlyResult: Option[Res] = None

  final def analyze(): Equation[Key, Value] = {
    pending.push(ProceedState(createStartState()))

    while (pending.nonEmpty && earlyResult.isEmpty) pending.pop() match {
      case MakeResult(state, delta, subIndices) =>
        val result = combineResults(delta, subIndices.map(results))
        if (isEarlyResult(result)) {
          earlyResult = Some(result)
        } else {
          val insnIndex = state.insnIndex
          results = results + (state.index -> result)
          computed = computed.updated(insnIndex, state :: computed(insnIndex))
        }
      case ProceedState(state) =>
        val insnIndex = state.insnIndex
        val shared = richControlFlow.isSharedInstruction(insnIndex)
        val conf = state.conf
        val history = state.history

        val loopEnter = dfsTree.loopEnters(insnIndex)
        val fold = loopEnter && history.exists(prevConf => confInstance(conf, prevConf))

        if (fold) {
          results = results + (state.index -> identity)
          if (shared)
            computed = computed.updated(insnIndex, state :: computed(insnIndex))
        } else computed(insnIndex).find(prevState => stateEquiv(state, prevState)) match {
          case Some(ps) =>
            Counter.effectivelyShared += 1
            val insnNode = methodNode.instructions.get(insnIndex)
            results = results + (state.index -> results(ps.index))
          case None =>
            processState(state)
        }
    }

    mkEquation(earlyResult.getOrElse(results(0)))
  }

  final def createStartFrame(): Frame[BasicValue] = {
    val frame = new Frame[BasicValue](methodNode.maxLocals, methodNode.maxStack)
    val returnType = Type.getReturnType(methodNode.desc)
    val returnValue = if (returnType == Type.VOID_TYPE) null else new BasicValue(returnType)
    frame.setReturn(returnValue)

    val args = Type.getArgumentTypes(methodNode.desc)
    var local = 0
    if ((methodNode.access & Opcodes.ACC_STATIC) == 0) {
      val basicValue = new NotNullValue(Type.getObjectType(controlFlow.className))
      frame.setLocal(local, basicValue)
      local += 1
    }
    for (i <- 0 until args.size) {
      val value = direction match {
        case InOut(`i`, _) =>
          new ParamValue(args(i))
        case In(`i`) =>
          new ParamValue(args(i))
        case _ =>
          new BasicValue(args(i))
      }
      frame.setLocal(local, value)
      local += 1
      if (args(i).getSize == 2) {
        frame.setLocal(local, BasicValue.UNINITIALIZED_VALUE)
        local += 1
      }
    }
    while (local < methodNode.maxLocals) {
      frame.setLocal(local, BasicValue.UNINITIALIZED_VALUE)
      local += 1
    }
    frame
  }

  final def popValue(frame: Frame[BasicValue]): BasicValue =
    frame.getStack(frame.getStackSize - 1)
}

object Utils {
  def isInstance(curr: Conf, prev: Conf): Boolean = {
    if (curr.insnIndex != prev.insnIndex) {
      return false
    }
    val currFr = curr.frame
    val prevFr = prev.frame
    for (i <- 0 until currFr.getLocals if !isInstance(currFr.getLocal(i), prevFr.getLocal(i)))
      return false
    for (i <- 0 until currFr.getStackSize if !isInstance(currFr.getStack(i), prevFr.getStack(i)))
      return false
    true
  }

  def isInstance(curr: BasicValue, prev: BasicValue): Boolean = prev match {
    case (_: ParamValue) => curr match {
      case _: ParamValue => true
      case _ => false
    }
    case InstanceOfCheckValue() => curr match {
      case InstanceOfCheckValue() => true
      case _ => false
    }
    case TrueValue() => curr match {
      case TrueValue() => true
      case _ => false
    }
    case FalseValue() => curr match {
      case FalseValue() => true
      case _ => false
    }
    case NullValue() => curr match {
      case NullValue() => true
      case _ => false
    }
    case NotNullValue(_) => curr match {
      case NotNullValue(_) => true
      case _ => false
    }
    case CallResultValue(_, prevInters) => curr match {
      case CallResultValue(_, currInters) => currInters == prevInters
      case _ => false
    }
    case _: BasicValue => true
  }

  def equiv(curr: Conf, prev: Conf): Boolean = {
    if (curr.insnIndex != prev.insnIndex) {
      return false
    }
    val currFr = curr.frame
    val prevFr = prev.frame
    for (i <- 0 until currFr.getLocals if !equiv(currFr.getLocal(i), prevFr.getLocal(i)))
      return false
    for (i <- 0 until currFr.getStackSize if !equiv(currFr.getStack(i), prevFr.getStack(i)))
      return false
    true
  }

  def equiv(curr: BasicValue, prev: BasicValue): Boolean = {
    val result = equiv0(curr, prev)
    val result2 = equiv0(prev, curr)
    require(result == result2, s"$result != $result2 || $curr (${curr.getClass}}), $prev (${prev.getClass}})")
    result
  }

  def equiv0(curr: BasicValue, prev: BasicValue): Boolean =
    if (curr.getClass == prev.getClass) {
      (curr, prev) match {
        case (CallResultValue(_, k1), CallResultValue(_, k2)) =>
          k1 == k2
        case _ =>
          true
      }
    }
    else false

}
